src/main/scala/fi/oph/kouta/indexing/Indexing.scala

```diff
@@ -0,0 +1,48 @@
+package fi.oph.kouta.indexing
+
+import fi.oph.kouta.domain._
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods.{compact, render}
+
+trait Indexing[A] {
+  def index: String
+  def indexId(a: A): Option[String]
+}
+
+trait WithOidIndexing[A <: PerustiedotWithOid] {
+  this: Indexing[A] =>
+
+  override def indexId(a: A): Option[String] = a.oid map { _.toString }
+}
+
+trait WithIdIndexing[A <: PerustiedotWithId] {
+  this: Indexing[A] =>
+
+  override def indexId(a: A): Option[String] = a.id map { _.toString }
+}
+
+object Indexing {
+  def apply[A](implicit in: Indexing[A]): Indexing[A] = in
+
+  object ops {
+    def indexMessage[A: Indexing](a: A): Option[String] = {
+      Indexing[A].indexId(a) map { oid => compact(render(Indexing[A].index -> Seq(oid))) }
+    }
+  }
+
+  implicit val hakuIndexing: Indexing[Haku] = new Indexing[Haku] with WithOidIndexing[Haku] {
+    val index: String = "haut"
+  }
```

Comment:
Olisi nätimpää, jos kutsuva service kertoo, mitä tyyppiä A se haluaa indeksoida.

Reply:
Niinhän se kertookin? Tää `Indexing[Haku]` on se joka kertoo mitä tarkkaan ottaen tehdään kun joku kutsuu `index(haku)`, vai tarkoitatko jotenkin et haku instanssilla pitäis voida indeksoija johonkin muuhun indeksiin?